src/python/pants/backend/python/util_rules/local_dists.py

```diff
@@ -0,0 +1,168 @@
+    EMPTY_SNAPSHOT,
+    Digest,
+    DigestContents,
+    DigestSubset,
+    MergeDigests,
+    PathGlobs,
+    Snapshot,
+)
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import TransitiveTargets, TransitiveTargetsRequest
+from pants.util.meta import frozen_after_init
+
+logger = logging.getLogger(__name__)
+
+
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class LocalDistsPexRequest:
+    """Request to build the local dists from the dependency closure of a set of addresses."""
+
+    addresses: Addresses
+    interpreter_constraints: InterpreterConstraints
+    # The result will return these with the sources provided by the dists subtracted out.
+    # This will help the caller prevent sources from appearing twice on sys.path.
+    sources: PythonSourceFiles
+
+    def __init__(
+        self,
+        addresses: Iterable[Address],
+        *,
+        interpreter_constraints: InterpreterConstraints = InterpreterConstraints(),
+        sources: PythonSourceFiles = PythonSourceFiles(
+            SourceFiles(EMPTY_SNAPSHOT, tuple()), tuple()
+        ),
+    ) -> None:
+        self.addresses = Addresses(addresses)
+        self.interpreter_constraints = interpreter_constraints
+        self.sources = sources
+
+
+@dataclass(frozen=True)
+class LocalDistsPex:
+    """A PEX file containing locally-built dists.
+
+    Can be consumed from another PEX, e.g., by adding to PEX_PATH.
+
+    Lists the files provided by the dists on sys.path, so they can be subtracted from
+    sources digests, to prevent the same file ending up on sys.path twice.
+    """
+
+    pex: Pex
+    # The sources from the request, but with any files provided by the local dists subtracted out.
+    remaining_sources: PythonSourceFiles
+
+
+@rule(desc="Building local distributions")
+async def build_local_dists(
+    request: LocalDistsPexRequest,
+) -> LocalDistsPex:
+
+    transitive_targets = await Get(TransitiveTargets, TransitiveTargetsRequest(request.addresses))
+
+    python_dist_field_sets = [
+        PythonDistributionFieldSet.create(target)
+        for target in transitive_targets.closure
+        if PythonDistributionFieldSet.is_applicable(target)
+    ]
+
+    dists = await MultiGet(
+        [Get(BuiltPackage, PackageFieldSet, field_set) for field_set in python_dist_field_sets]
+    )
+
+    # The primary use-case of the "local dists" feature is to support consuming native extensions
+    # as wheels without having to publish them first.
+    # It doesn't seem very useful to consume locally-built sdists, and it makes it hard to
+    # reason about possible sys.path collisions between the in-repo sources and whatever the
+    # sdist will place on the sys.path when it's installed.
+    # So for now we simply ignore sdists, with a warning if necessary.
+    provided_files = set()
+    wheels = []
+
+    for dist in dists:
+        contents = await Get(DigestContents, Digest, dist.digest)
```

Comment:
Should be a `MultiGet` and then you use `for dist, contents in zip(dists, all_contents)`